### DISCUSSION TITLE
Seeking help on how to obtain synchronized conversations.

### DISCUSSION THREAD

**Discussion by super-zzc** [tulir/whatsmeow#692] [WhatsApp protocol Q&A] [answered]
I am seeking help because I don't know how to retrieve the conversation list. Using client.Store.Contacts.GetAllContacts() and client.GetJoinedGroups() only allows me to get the list of added contacts and groups, but I am unable to obtain conversations when I haven't added the contact and yet have a conversation with them. I'm not sure how to get the synchronized conversation list.

**Comment by masewo** [comment]
As far as I know there is no interface like that. You have to create and hold conversations on your application side. You can reconstruct them while parsing all incoming messages (either coming from history sync or live).

Keep in mind that newly created conversations from your application side have to be threated separately.

**Reply by super-zzc** [reply]
> 据我所知，没有这样的界面。您必须在应用程序端创建并保持对话。您可以在解析所有传入消息（来自历史同步或实时消息）时重建它们。
> 
> 请记住，从应用程序端新创建的对话必须受到单独威胁。

Hi, thank you for your response. On the web version of WhatsApp and the mobile app, I can initiate conversations with people who are not added as contacts without any issues. I'm wondering if whatsmeow supports retrieving such conversation lists. Are you saying that I must engage in a conversation with the other party using whatsmeow and then rely on the historical messages to recreate it?

**Reply by masewo** [reply]
Yes, you can create conversations with people that are not in your contacts. But maybe you should check if they are on WhatsApp first. 

If you connect your session the first time you get all messages synced and should store them.

**Reply by super-zzc** [reply]
> 序上，我可以毫无问题地与未添加为联系人的人发起对话。我想知道 whatsmeow 是否支持检索此类对话列表。您是说我必须使用

Hello, I need to maintain my login on whatsmeow to receive conversations from non-contacts either when they initiate the conversation with me or when I initiate it with them. For past conversations with non-contacts, I am unable to retrieve them, correct? It is only after one party initiates the conversation that I can then obtain the chat history.

**Reply by masewo** [reply, marked as answer]
There is no difference if it is a contact or a non-contact.

The flow is like this:
The first time after you scan the QR code to login (or type in the login code), you will receive all past messages in the history sync events. You need to save them in your own database.

While you save them, it's good practice to save them in a message table and in a conversation table. For every first message of a distinct JID you create a new conversation.

**Reply by andresovich** [reply]
I am looking into requesting the history of all chats again after the qr code has been scanned. Is there a way to retrieve this info from whatsapp after the scann of the code?

Do you know what happens if the connection gets interrupted in the middle of the syncing process?
Thanks! great work!

**Comment by SaCavid** [comment]
Just to speed up things. We are receiving *events.HistorySync only after login just once. Here is sync part in my code.


```
func (h *Hub) eventHandler(evt interface{}) func(evt interface{}) {
	return func(evt interface{}) {
		switch v := evt.(type) {
		case *events.Message:
			log.Println("Received whatsapp message: ", v.Info.Chat.String(), "-", v.Info.Sender.String(), "-", v.Info.Type)
			convo := v.Message.GetConversation()
			if convo != "" {
				payload := MessagePayload{
					From:    v.Info.Chat.String(),
					Message: convo,
					Date:    getCurrentTime(),
					Type:    "incoming",
				}
				h.broadcast <- payload
			}
		case *events.HistorySync:
			fmt.Println("Received a history sync")
			for _, conv := range v.Data.GetConversations() {
				for _, historymsg := range conv.GetMessages() {
					chatJID, _ := types.ParseJID(conv.GetID())
					evt, err := h.wClient.ParseWebMessage(chatJID, historymsg.GetMessage())
					if err != nil {
						log.Println(err)
					}
					fmt.Println("Received a history sync")
					
					convo := evt.Message.GetConversation()
					if convo != "" {
						payload := MessagePayload{
							From:    evt.Info.Chat.String(),
							Message: convo,
							Date:    getTimeFromTimeStamp(v.Info.Timestamp),
							Type:    "incoming", // internal type
						}
						h.broadcast <- payload
					}
				}
			}
		}
	}
}```

**Reply by vitorvk2** [reply]
Thanks!!!

**Reply by Dan1718** [reply]
Thank you!
